core/src/nucleus/ribosome/api/capabilities.rs

```diff
@@ -0,0 +1,83 @@
+use crate::{
+    agent::actions::commit::commit_entry,
+    nucleus::ribosome::{api::ZomeApiResult, Runtime},
+};
+use holochain_core_types::{
+    cas::content::Address,
+    entry::{cap_entries::CapTokenGrant, Entry},
+    error::HolochainError,
+};
+use holochain_wasm_utils::api_serialization::capabilities::GrantCapabilityArgs;
+use std::convert::TryFrom;
+use wasmi::{RuntimeArgs, RuntimeValue};
+
+pub fn invoke_grant_capability(runtime: &mut Runtime, args: &RuntimeArgs) -> ZomeApiResult {
+    let context = runtime.context()?;
+    // deserialize args
+    let args_str = runtime.load_json_string_from_args(&args);
+    let args = match GrantCapabilityArgs::try_from(args_str.clone()) {
+        Ok(input) => input,
+        Err(..) => return ribosome_error_code!(ArgumentDeserializationFailed),
+    };
+
+    let task_result: Result<Address, HolochainError> =
+        match CapTokenGrant::create(&args.id, args.cap_type, args.assignees, args.functions) {
+            Ok(grant) => context.block_on(commit_entry(
```

Comment:
This is another example where I think we could compose with futures instead. You are essentially forced to block here because you must return a zome api result. It is outside the scope of the PR, but we could change all these api functions to return a future zome api result, and universally handle the unraveling of these futures at the last moment when serializing over the network. In this model no explicit blocking is ever really necessary- even during initialization, if this is done right. Instead you just transform an existing future which achieves the same behavior.

Reply:
Hmm interesting.  Adding an item to tech-debt to talk this through.